### ISSUE TITLE
Allow specifying apt-packages for subset of build matrix

### ISSUE THREAD

**Issue by cancan101** [travis-ci/travis-ci#3505]
I would like to run the apt-packages add-on only for a subet of my build matrix.

http://docs.travis-ci.com/user/apt-packages/

**Comment by BanzaiMan** [comment]
@cancan101 Could you elaborate on what issues you are having? APT package addon is processed by the build script after all the build matrix process has been completed. Making the change is not trivial.

**Comment by cancan101** [comment]
What I would like would be to specify what apt packages to install with which build. Right now my `.travis.yml` file looks like:
```
env:
 - SERVICE=s1
 - SERVICE=s2
 - SERVICE=s3
 - SERVICE=s4
...
addons:
  apt_packages:
    - nginx
```
but I want to install `nginx` only for some of the `env` values.

**Comment by ldionne** [comment]
I need this too. My situation is that I need to install one of

    clang-3.5
    clang-3.6
    clang-3.7

depending on the CLANG_VERSION environment variable defined in my build matrix. My initial solution was to simply install them all and then use only the one I need. However, this is not possible since clang-3.6 can not be installed in parallel with clang-3.5.

So the problem is not only that it is superfluous to install the 3 packages at the same time even when only one is needed, but it is sometimes straight impossible to do so.

One possible fix would be to set the environment variables before the addons are performed, which would allow me to write

    addons:
        apt_packages:
            - clang-${CLANG_VERSION}


It is not the most generic fix as I do not see how it solves the OP's problem, but it would solve mine.

**Comment by joshk** [comment]
Thanks for the suggestion.

I'm sorry but we aren't able to export env vars first as this is a security risk as someone could embed some shell which would use sudo. (sudo is a security risk on the container based infrastructure).

We have some changes in the works which may allow for something like this in the form of specifying your exact configs for each jobs, including addons. 

We don't have an ETA for this though.

**Comment by cancan101** [comment]
There seems to be a trends toward monorepos such as Google and [now Facebook](http://gregoryszorc.com/blog/2015/03/28/notes-from-facebook%27s-developer-infrastructure-at-scale-f8-talk/) so having the ability to specify different languages within one repos is important.

**Comment by andyli** [comment]
Turn out I've just discovered that, we can already able to do similar thing with `matrix: include:`:

```yaml
language: haxe
matrix:
  include:
    - haxe: "3.1.3"
      addons:
        apt:
          packages:
            - php-zeroc-ice # use it to install php5 for now
      install:
        - yes | haxelib install hxcpp
        - yes | haxelib install hxjava
        - yes | haxelib install hxcs
        - haxelib list

    - haxe: "development"
      addons:
        apt:
          packages:
            - php-zeroc-ice # use it to install php5 for now
            - python3
      install:
        # install and rebuild hxcpp
        - haxelib git hxcpp https://github.com/HaxeFoundation/hxcpp.git
        - cd $(haxelib path hxcpp | head -1)tools/hxcpp && haxe compile.hxml
        - cd $(haxelib path hxcpp | head -1)project && neko build.n linux-m64
        - cd $TRAVIS_BUILD_DIR

        # install hxjava and hxcs
        - haxelib git hxjava https://github.com/HaxeFoundation/hxjava.git
        - haxelib git hxcs https://github.com/HaxeFoundation/hxcs.git

        - haxelib list
```

Although in this way the matrix has to be expanded manually with potentially lots of duplicated parts.

**Comment by cancan101** [comment]
[yaml supports references](https://en.wikipedia.org/wiki/YAML#Advanced_components_of_YAML). I wonder if those can be used here?

**Comment by cancan101** [comment]
```yml
language: haxe
matrix:
  include:
    - haxe: "3.1.3"
      addons:
        apt:
          packages:
            - php-zeroc-ice
      install: &install
        - yes | haxelib install hxcpp

    - haxe: "development"
      addons:
        apt:
          packages:
            - php-zeroc-ice
            - python3
      install: *install
```

**Comment by weitjong** [comment]
Currently both the ```addons.apt.sources``` and ```addons.apt.packages``` keys only accept a list of values. Could it be enhanced to also accept a list of list of values? For example:
```
addons:
  apt:
    packages:
      - packageA    # Normal list of values
      - packageB
      - [ packageC, packageD ]   # List of list
      - [ packageE, packageF ]
```

If this is possible then one could use yaml anchor and alias to mix and match subset of lists easily in the job matrix. For example:
```
addons:
  apt:
    packages: &native_deps
      - packageA
      - packageB

matrix:
  include:
    # Job 1
    - env: 64BIT_NATIVE_BUILD=1
      addons:
        apt:
          packages:
            - *native_deps
            - [packageC, packageD]    # 64-bit specific packages goes here
    # Job 2
    - env: 32BIT_NATIVE_BUILD=1
      addons:
        apt:
          packages:
            - *native_deps
            - &32bit_deps [packageF:i386, packageG:i386]    # 32-bit specific goes here
    # Job 3
    - env: 32BIT_MINGW_BUILD=1
      addons:
        apt:
          packages:
            - [packageH, packageI]   # MinGW specific packages goes here
            - *32bit_deps  # Reuse the 32-bit dependency list one more time
```

This is just an imaginary scenario to illustrate the possibility.

**Comment by meatballhat** [comment]
@weitjong This is a great suggestion IMHO.  Combining `matrix.include`, yaml refs, and nested arrays looks like it might cover everything.  We're also looking to improve the UX around the operations possible with `matrix.include` but haven't settled on a solution just yet.

**Comment by weitjong** [comment]
@meatballhat That was fast! I am really impressed. However, I have just tested this new enhancement and found that although the nested list is being parsed and accepted now, it did not pass the whitelist check. This is what I have tested in my quick test.
```
addons: {apt: {packages: [[libasound2-dev], [graphviz], [rpm, doxygen]]}}
```
And this is the result.
```
Installing APT Packages (BETA)
Disallowing packages: libasound2-dev, graphviz, rpm, doxygen
If you require these packages, please review the package approval process at: https://github.com/travis-ci/apt-package-whitelist#package-approval-process
```
All those packages are in the whitelist. Manually flatten the list works as before. So, I suppose somewhere else need to be enhanced as well to handle the nested array.

**Comment by meatballhat** [comment]
@weitjong It's not quite live yet! :smiley:  I've merged and deployed to our staging instance, but need to wait until morning to deploy to production.

**Comment by weitjong** [comment]
I see. Stupid me. :smiley:

**Comment by meatballhat** [comment]
@weitjong not at all!  I meant to say as much :smiley_cat:

**Comment by cancan101** [comment]
Any reason to write it as in https://github.com/travis-ci/travis-ci/issues/3505#issuecomment-92764855 but without having to put `*native_deps` in each matrix entry? ie have the items in the global `addons::apt` be inherited?

**Comment by weitjong** [comment]
@cancan101 In my hypothetical example, the list are slightly different for each job. If, however, the dep list of the job is identical to global addons then there is no need to overwrite it and it should be inherited.

**Comment by andyli** [comment]
Could nested list be supported for other keys, e.g. `install:` and `script:`?

**Comment by cancan101** [comment]
@weitjong I was thinking of the case where there is some set of global requirements and then each of the different builds has some additional specific requirements.

**Comment by weitjong** [comment]
@cancan101 I think I understand you. It just that yaml does not support inheritance in the OOP sense. This is how I understand how yaml works. We can define a node (map or list) with an anchor so that it can be reused later with an alias. In the "different builds" we can overwrite what have been defined globally. In this case, the overwrite can reuse the node aliases as they fit the requirement. The node alias from "global" or some other part of the yaml document is just mimicking the inheritance behavior.

**Comment by cancan101** [comment]
Right this merging of a global list and a build specific list would have to be done by travis-ci.

**Comment by weitjong** [comment]
If this is done at Travis CI side then it needs to work out all use cases including the case when a local ```matrix.include``` job does not want to inherit any of the global things (addons, env, or others).

**Comment by meatballhat** [comment]
This is now live in production :metal: :smile_cat: :heart:

**Comment by weitjong** [comment]
Confirmed it is working now. Thank you.

**Comment by thedrow** [comment]
Any idea why this is not working: https://github.com/thedrow/rapidjson/blob/caa4d22d1680f22e26c07f3a2c47354f121479da/.travis.yml
Travis won't even start the build.
Also, can we please add documentation for this one.

**Comment by BanzaiMan** [comment]
@thedrow It appears that we don't know about this repository yet. Is it a recent fork? Try resyncing data, and enabling it. You'd need to push a new commit to start building.

**Comment by thedrow** [comment]
There's a pull request for it but the build wasn't triggered.

**Comment by BanzaiMan** [comment]
@thedrow You mean you opened a PR to miloyip/rapidjson and it didn't fire? https://travis-ci.org/miloyip/rapidjson/requests shows that PR for https://github.com/miloyip/rapidjson/pull/504 is "missing config", which is sometimes an indication that `.travis.yml` is invalid YAML. Indeed, https://github.com/thedrow/rapidjson/blob/caa4d22d1680f22e26c07f3a2c47354f121479da/.travis.yml#L29 has a `:` in the far-out place where the YAML parser will try to interpret it as a delimiter. (There is a similar issue on line 32.) You should split that up into multiple lines:

```yaml
matrix:
  include:
    env: CONF=debug   ARCH=x86
    addons:
      apt:
        packages:
          - *default_packages
          - g++-multilib
          - libc6-dbg:i386
⋮
```

**Comment by thedrow** [comment]
Thanks!

**Comment by thedrow** [comment]
It seems that this doesn't install new packages for the existing jobs in the matrix.

**Comment by thedrow** [comment]
And the plot thickens. I tried to exclude builds and got this instead: https://travis-ci.org/miloyip/rapidjson/builds/99414655
Should I open another issue?

**Comment by BanzaiMan** [comment]
@thedrow That is a completely separate issue, though. I suggest opening a separate issue. I don't know how CMake and clang (and their various versions) interact with each other; it might be that Precise is just too old to handle these things.
